File: DS/slidingWindow/newDistinctNumber.java
import java.util.HashMap;

// javac avgSW.java
// java avgSW


public class newDistinctNumber{

    public static int distinctNumber(int[] arr, int k){

        int maxLength =0;
        int left =0;

        HashMap<Integer, Integer> counts = new HashMap<>();

        for (int right= 0; right< arr.length; right ++){

            int currentNum = arr[right];

            Integer currentCount = counts.get(currentNum);

            if(currentCount == null){
                counts.put(currentNum,1);
            }else{
                counts.put(currentNum, currentCount + 1);

            }


            while(counts.size() > k){

                int numberToExit = arr[left];

                Integer countAtLeft = counts.get(numberToExit);
                if(countAtLeft > 1){
                    counts.put(numberToExit, countAtLeft - 1 );

                }else{
                    counts.remove(numberToExit);

                }

                left = left +1;

            }

            int currentWindowSize = right-left + 1;
        if(currentWindowSize > maxLength){
            maxLength = currentWindowSize;
        }

        }

        return maxLength;


        
    }

    public static void main(String[] args){

        int[] arr ={1, 2, 1, 2, 3};
        int k =2;

        System.out.println(distinctNumber(arr, k));

    }

}

File: DS/slidingWindow/newDistinctNumber.js
// Problem

// Given an integer array nums and an integer k, return the length of the longest subarray that contains at most k distinct numbers.

// nums = [1, 2, 1, 2, 3]
// k = 2

// output : 4



// we added one item to the map, but still why second time also currentCount shows undefined, can you explain like you explain to a beginner

// currentCount---> undefined
// inside if counts.set(currentNum,1)---> Map(1) { 1 => 1 }
// currentCount---> undefined
// inside if counts.set(currentNum,1)---> Map(2) { 1 => 1, 2 => 1 }
// currentCount---> 1
// inside else counts.set(currentNum, currentCount + 1)---> Map(2) { 1 => 2, 2 => 1 }
// currentCount---> 1
// inside else counts.set(currentNum, currentCount + 1)---> Map(2) { 1 => 2, 2 => 2 }
// currentCount---> undefined
// inside if counts.set(currentNum,1)---> Map(3) { 1 => 2, 2 => 2, 3 => 1 }

function newDistinctNumber(arr, K){

    let maxLength = 0;
    let left=0;
    let counts = new Map();

    for(let right = 0; right< arr.length; right++){
        let currentNum = arr[right];

        let currentCount = counts.get(currentNum);

        console.log("currentCount--->", currentCount);

        if(currentCount == undefined){
            console.log("inside if currentNum--->", currentNum);
            counts.set(currentNum,1);
            console.log("inside if counts.set(currentNum,1)--->", counts.set(currentNum,1));

        }else{
            console.log("inside else currentNum--->", currentNum , "=>", currentCount + 1 );

            counts.set(currentNum, currentCount + 1);
            console.log("inside else counts.set(currentNum, currentCount + 1)--->", counts.set(currentNum, currentCount + 1));

        }

        console.log("counts.size --->", counts.size);

        while(counts.size > K){
            let numberToExit = arr[left];

            let currentCount = counts.get(numberToExit);
                    console.log("inside while currentCount --->", currentCount);

            if(currentCount >1){
                counts.set(numberToExit,  currentCount - 1 );

            }
            else{
                counts.delete(numberToExit);

            }



            left = left + 1;

        }

        let currentWindowSize = right-left + 1;
        if(currentWindowSize > maxLength){
            maxLength = currentWindowSize;
        }

    }

    return maxLength;


}

console.log(newDistinctNumber([1, 2, 1, 2, 3],2));


